service/matching/matching_engine.go

```diff
@@ -837,6 +837,155 @@ func (e *matchingEngineImpl) listTaskQueuePartitions(request *matchingservice.Li
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_InsertAssignmentRule:
+			versioningData, err = InsertAssignmentRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetInsertAssignmentRule(),
+				e.config.VersionAssignmentRuleLimitPerQueue(ns.Name().String()),
+			)
+			if err != nil {
+				// operation can't be completed due to limits. no action, do not replicate, report error
+				return nil, false, err
+			}
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_ReplaceAssignmentRule:
+			versioningData, err = ReplaceAssignmentRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetReplaceAssignmentRule(),
+			)
+			if err != nil {
+				// operation can't be completed due to limits. no action, do not replicate, report error
+				return nil, false, err
+			}
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_DeleteAssignmentRule:
+			versioningData, err = DeleteAssignmentRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetDeleteAssignmentRule(),
+			)
+			if err != nil {
+				// operation can't be completed due to limits. no action, do not replicate, report error
+				return nil, false, err
+			}
+			if ns.ReplicationPolicy() == namespace.ReplicationPolicyMultiCluster {
+				operationCreatedTombstones = true
+			} else { // todo: figure out if this is the right thing to do with tombstones
+				// We don't need to keep the tombstones around if we're not replicating them.
+				versioningData = ClearTombstones(versioningData)
+			}
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_InsertCompatibleRedirectRule:
+			versioningData, err = InsertCompatibleRedirectRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetInsertCompatibleRedirectRule(),
+				e.config.VersionCompatibleRedirectRuleLimitPerQueue(ns.Name().String()),
+			)
+			if err != nil {
+				// operation can't be completed due to limits. no action, do not replicate, report error
+				return nil, false, err
+			}
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_ReplaceCompatibleRedirectRule:
+			versioningData, err = ReplaceCompatibleRedirectRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetReplaceCompatibleRedirectRule(),
+			)
+			if err != nil {
+				// operation can't be completed due to limits. no action, do not replicate, report error
+				return nil, false, err
+			}
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_DeleteCompatibleRedirectRule:
+			versioningData, err = DeleteCompatibleRedirectRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetDeleteCompatibleRedirectRule(),
+			)
+			if err != nil {
+				// operation can't be completed due to limits. no action, do not replicate, report error
+				return nil, false, err
+			}
+			if ns.ReplicationPolicy() == namespace.ReplicationPolicyMultiCluster {
+				operationCreatedTombstones = true
+			} else { // todo: figure out if this is the right thing to do with tombstones
```

Comment:
Let's add a config for retention of deleted rules. I think we can keep those rules for a week or so to help with troubleshooting. Later we may adjust the value.
For the behavior, I think we can call `ClearTombstones` at the end, regardless of the operation. Consider renaming it to `CleanupTombstones` because it does not clear all tombstones, only the ones old enough.